Navigation component: Add back button click handler

david-szabo97 opened the pull request:
Related: https://github.com/WordPress/gutenberg/issues/25247

## Description
Allow specifying custom back button click handler.

## How has this been tested?
- Run in terminal: `yarn storybook:dev`
- Click **Category**
- Click **Custom back click handler**
- Click **Go back to root**

## Screenshots
![image](https://user-images.githubusercontent.com/2256104/94013825-ed8e1200-fdaa-11ea-9f88-f54520e819be.png)

## Types of changes
Enhancement

## Checklist:
- [x] My code is tested.
- [x] My code follows the WordPress code style. <!-- Check code: `npm run lint`, Guidelines: https://make.wordpress.org/core/handbook/best-practices/coding-standards/javascript/ -->
- [ ] My code follows the accessibility standards. <!-- Guidelines: https://make.wordpress.org/core/handbook/best-practices/coding-standards/accessibility-coding-standards/ -->
- [ ] My code has proper inline documentation. <!-- Guidelines: https://make.wordpress.org/core/handbook/best-practices/inline-documentation-standards/javascript/ -->
- [ ] I've included developer documentation if appropriate. <!-- Handbook: https://developer.wordpress.org/block-editor/ -->
- [ ] I've updated all React Native files affected by any refactorings/renamings in this PR. <!-- React Native mobile Gutenberg guidelines: https://github.com/WordPress/gutenberg/blob/master/docs/contributors/native-mobile.md -->

david-szabo97 committed:
Add onBackButtonClick prop

david-szabo97 committed:
Fix onClick not being called

david-szabo97 committed:
Simplify tertiary

david-szabo97 commented:
Back button will be shown if either `parentMenu` or `onBackButtonClick` prop is specified on `NavigationMenu`

I wonder which should take priority. Right now if both `parentMenu` and `onClick` are specified in the `NavigationBackButton` then the `parentMenu` takes priority so the button will act as a back button to the parent menu.

It doesn't make sense to have both props specified, but `parentMenu` might act as an information. For example, when I'm creating menus it would be great to see what's the parent menu of the menu I'm creating. But I'd like to have a custom back button click handler.

IMO, `onClick` should take precedence. Or we can utilise the `event.preventDefault()`.

```js
<NavigationMenu
	menu="custom-back-click"
	title="Custom back button click handler"
	onBackButtonClick={ ( e ) => {
		e.preventDefault(); // This will prevent the default onClick handler which navigates to the parentMenu
		setActiveMenu( 'root' );
	} }
	backButtonLabel="Go back to root"
/>
```

david-szabo97 committed:
Update readme

mattwiebe reviewed (approved):
Tests well, code looks good

vindl commented on packages/components/src/navigation/stories/index.js:
```diff
@@ -147,6 +147,11 @@ function Example() {
 					/>
+					<NavigationItem
+						navigateToMenu="custom-back-click"
+						item="child-5"
```
This should be `child-6`.

vindl reviewed (commented):
Overall this looks good but needs rebase after latest Storybook reorganization from @Copons.

vindl commented:
> I wonder which should take priority. Right now if both parentMenu and onClick are specified in the NavigationBackButton then the parentMenu takes priority so the button will act as a back button to the parent menu.
...
IMO, onClick should take precedence. Or we can utilise the event.preventDefault().

I also think that onClick should take precedence but not completely override it, since we might want to execute some side effects in it, and then continue with back navigation as usual (e.g. let's say someone wants to track stats for back button clicks in their component or similar). On the other hand, if someone wants a complete override of the behavior, they can optionally use `event.preventDefault()` as you've pointed out.

david-szabo97 committed:
Allow prevention of navigating back

david-szabo97 committed:
Add stories

david-szabo97 commented:
* `onBackButtonClick` will be called with the `Event` object. This way we can use `event.preventDefault()` to prevent the default behavior (navigate back to `parentMenu`)
* Rebased and added stories.

noahtallen commented on packages/components/src/navigation/back-button/index.js:
```diff
@@ -29,15 +29,23 @@ function NavigationBackButton(
 
 	const parentMenuTitle = navigationTree.getMenu( parentMenu )?.title;
 
+	const handleOnClick = ( event ) => {
```
Does it make sense to add `useCallback` here to maintain a single reference unless a dependency changes?

david-szabo97 replied:
I didn't profile it, but I think `useCallback` would make this worse. This function is simple and it is only used in one component, so we would only save 1 rerender which is probably cheaper than `useCallback`. 🤔

noahtallen replied:
Sounds good!

Copons commented on packages/components/src/navigation/stories/more-examples.js:
```diff
@@ -83,6 +98,33 @@ export function MoreExamplesStory() {
+					title="You can't go back from here"
+					badge={ backButtonPreventedBadge }
+				/>
+			</NavigationMenu>
```
Ahaha this example is EVIL!

I like it.
